> I think `linuxPackages` defaults to 6.6.

Ah I see what you mean, sorry. Yep, it fails building on 6.6.30:
<details>
<summary><code> nix build -L .#linuxPackages.ena
ena> Running phase: unpackPhase
ena> unpacking source archive /nix/store/7nnir1y4wikh13mrhrpdz0zk6ap9m4rn-source
ena> source root is source
ena> Running phase: patchPhase
ena> Running phase: updateAutotoolsGnuConfigScriptsPhase
ena> Running phase: configurePhase
ena> substituteStream(): WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. (file 'Makefile')
ena> Running phase: buildPhase
ena> build flags: SHELL=/nix/store/h3bhzvz9ipglcybbcvkxvm4vg9lwvqg4-bash-5.2p26/bin/bash O=\$\(buildRoot\) CC=/nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc HOSTCC=/nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc HOSTLD=/nix/store/m3gq2f6iwh3g67zrs27m7w82b37l8gf8-binutils-wrapper-2.41/bin/ld ARCH=x86_64
ena> ./configure.sh --kernel-dir /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build --verbose
ena> Trying to intercept CC arguments
ena> Detected compiler:  /nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc
ena> Managed to intercept CC arguments
ena> Command: /nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin/cc -nostdinc -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/arch/x86/include -I./arch/x86/include/generated -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include -I./include -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/uapi -I./include/generated/uapi -include /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/linux/compiler-version.h -include /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/linux/kconfig.h -include /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/include/linux/compiler_types.h -D__KERNEL__ -fmacro-prefix-map=/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/= -std=gnu11 -fshort-wchar -funsigned-char -fno-common -fno-PIE -fno-strict-aliasing -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -fcf-protection=branch -fno-jump-tables -m64 -falign-jumps=1 -falign-loops=1 -mno-80387 -mno-fp-ret-in-387 -mpreferred-stack-boundary=3 -mskip-rax-setup -mtune=generic -mno-red-zone -mcmodel=kernel -fno-asynchronous-unwind-tables -mindirect-branch=thunk-extern -mindirect-branch-register -mindirect-branch-cs-prefix -mfunction-return=thunk-extern -fno-jump-tables -mharden-sls=all -fpatchable-function-entry=16,16 -fno-delete-null-pointer-checks -O2 -fno-allow-store-data-races -fstack-protector-strong -ftrivial-auto-var-init=zero -fno-stack-clash-protection -pg -mrecord-mcount -mfentry -DCC_USING_FENTRY -falign-functions=16 -fstrict-flex-arrays=3 -fno-strict-overflow -fno-stack-check -fconserve-stack -g -DMODULE -DKBUILD_BASENAME="main" -DKBUILD_MODNAME="main" -D__KBUILD_MODNAME=kmod_main -fno-pie -Werror=incompatible-pointer-types -Wno-nonnull -Werror=implicit -isystem /nix/store/9hgsinpfgyvsd92v0wlvmxv9wnaal68r-gcc-13.2.0/lib/gcc/x86_64-unknown-linux-gnu/13.2.0/include
ena> ### Detection of the available kernel API ###
ena> ENA_HAVE_PCI_DEV_ID: YES
ena> ENA_HAVE_XDP_DO_FLUSH: YES
ena> ENA_HAVE_CPUMASK_LOCAL_SPREAD: YES
ena> ENA_HAVE_UPDATE_AFFINITY_HINT: YES
ena> ENA_HAVE_ETHTOOL_PUTS: YES
ena> ENA_HAVE_ETHTOOL_RXFH_PARAM: NO
ena> ### Detection completed ###
ena> make -C /nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build M=/build/source/kernel/linux/ena modules
ena> make[1]: Entering directory '/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build'
ena>   CC [M]  /build/source/kernel/linux/ena/ena_netdev.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_ethtool.o
ena> /build/source/kernel/linux/ena/ena_ethtool.c: In function 'ena_metrics_stats_strings':
ena> /build/source/kernel/linux/ena/ena_ethtool.c:439:33: warning: implicit declaration of function 'ethtool_puts'; did you mean 'ethtool_notify'? [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
ena>   439 |                                 ethtool_puts(data, ena_metrics->name);
ena>       |                                 ^~~~~~~~~~~~
ena>       |                                 ethtool_notify
ena>   CC [M]  /build/source/kernel/linux/ena/ena_lpc.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_phc.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_xdp.o
ena>   CC [M]  /build/source/kernel/linux/ena/dim.o
ena>   CC [M]  /build/source/kernel/linux/ena/net_dim.o
ena>   CC [M]  /build/source/kernel/linux/ena/../common/ena_com//ena_com.o
ena>   CC [M]  /build/source/kernel/linux/ena/../common/ena_com//ena_eth_com.o
ena>   CC [M]  /build/source/kernel/linux/ena/ena_sysfs.o
ena>   LD [M]  /build/source/kernel/linux/ena/ena.o
ena>   MODPOST /build/source/kernel/linux/ena/Module.symvers
ena> ERROR: modpost: "ethtool_puts" [/build/source/kernel/linux/ena/ena.ko] undefined!
ena> make[3]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/scripts/Makefile.modpost:145: /build/source/kernel/linux/ena/Module.symvers] Error 1
ena> make[2]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:1865: modpost] Error 2
ena> make[1]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:234: __sub-make] Error 2
ena> make[1]: Leaving directory '/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build'
ena> make: *** [Makefile:102: ena.ko] Error 2
error: builder for '/nix/store/blxa9qml79w7lrm0h6d2vw07l57w39l9-ena-2.12.0-6.6.30.drv' failed with exit code 2;
       last 10 log lines:
       >   CC [M]  /build/source/kernel/linux/ena/../common/ena_com//ena_eth_com.o
       >   CC [M]  /build/source/kernel/linux/ena/ena_sysfs.o
       >   LD [M]  /build/source/kernel/linux/ena/ena.o
       >   MODPOST /build/source/kernel/linux/ena/Module.symvers
       > ERROR: modpost: "ethtool_puts" [/build/source/kernel/linux/ena/ena.ko] undefined!
       > make[3]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/scripts/Makefile.modpost:145: /build/source/kernel/linux/ena/Module.symvers] Error 1
       > make[2]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:1865: modpost] Error 2
       > make[1]: *** [/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/source/Makefile:234: __sub-make] Error 2
       > make[1]: Leaving directory '/nix/store/w2bzm4kvp46xhsqpbqvp1yyxmp5zq0fz-linux-6.6.30-dev/lib/modules/6.6.30/build'
       > make: *** [Makefile:102: ena.ko] Error 2
       For full logs, run 'nix log /nix/store/blxa9qml79w7lrm0h6d2vw07l57w39l9-ena-2.12.0-6.6.30.drv'.</code></summary></details>

I'll check if there's any workaround.

--- a/pkgs/os-specific/linux/ena/default.nix
+++ b/pkgs/os-specific/linux/ena/default.nix
@@ -19,6 +19,12 @@ stdenv.mkDerivation rec {
   # linux 3.12
   env.NIX_CFLAGS_COMPILE = "-Wno-error=implicit-function-declaration";
 
+  patches = [
+    # Use kernel version checks instead of API feature detection
+    # See https://github.com/NixOS/nixpkgs/pull/310680
+    ./override-features-api-detection.patch
+  ];
+
   configurePhase = ''
     runHook preConfigure
     cd kernel/linux/ena

--- a/pkgs/os-specific/linux/ena/override-features-api-detection.patch
+++ b/pkgs/os-specific/linux/ena/override-features-api-detection.patch
@@ -1,0 +1,55 @@
+diff --git a/kernel/linux/ena/kcompat.h b/kernel/linux/ena/kcompat.h
+index 32a9cc5..8d39362 100644
+--- a/kernel/linux/ena/kcompat.h
++++ b/kernel/linux/ena/kcompat.h
+@@ -888,21 +888,6 @@ xdp_prepare_buff(struct xdp_buff *xdp, unsigned char *hard_start,
+ #define ENA_XDP_XMIT_FREES_FAILED_DESCS_INTERNALLY
+ #endif
+ 
+-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0) && \
+-	!(LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 188) && \
+-	 LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)) && \
+-	!(LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 251) && \
+-	 LINUX_VERSION_CODE < KERNEL_VERSION(5, 5, 0))) && \
+-	!(defined(RHEL_RELEASE_CODE) && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(8, 6)) && \
+-	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL >= 4)) && \
+-	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL == 3) && \
+-	  ENA_KERNEL_VERSION_GTE(5, 3, 18, 150300, 59, 43))
+-static inline void eth_hw_addr_set(struct net_device *dev, const u8 *addr)
+-{
+-	memcpy(dev->dev_addr, addr, ETH_ALEN);
+-}
+-#endif
+-
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0) || \
+ 	(defined(RHEL_RELEASE_CODE) && \
+ 	RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(8, 6) && \
+@@ -1112,7 +1097,7 @@ static inline void ena_dma_unmap_page_attrs(struct device *dev,
+ #define pci_dev_id(pdev) ((((u16)(pdev->bus->number)) << 8) | (pdev->devfn))
+ #endif /* ENA_HAVE_PCI_DEV_ID */
+ 
+-#ifndef ENA_HAVE_XDP_DO_FLUSH
++#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 6, 0)
+ #define xdp_do_flush xdp_do_flush_map
+ #endif /* ENA_HAVE_XDP_DO_FLUSH */
+ 
+@@ -1147,15 +1132,15 @@ static inline unsigned int cpumask_local_spread(unsigned int i, int node)
+ }
+ #endif /* ENA_HAVE_CPUMASK_LOCAL_SPREAD */
+ 
+-#ifndef ENA_HAVE_UPDATE_AFFINITY_HINT
++#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
+ static inline int irq_update_affinity_hint(unsigned int irq, const struct cpumask *m)
+ {
+ 	return 0;
+ }
+-#endif /* ENA_HAVE_UPDATE_AFFINITY_HINT */
++#endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5.17.0) */
+ 
+-#ifndef ENA_HAVE_ETHTOOL_PUTS
++#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0)
+ #define ethtool_puts ethtool_sprintf
+-#endif /* ENA_HAVE_ETHTOOL_PUTS */
++#endif
+ 
+ #endif /* _KCOMPAT_H_ */

